### PR TITLE
Feature/oracle agent spec integration

### PR DESCRIPTION
**Pull request by afourniernv**
## Overview

This PR represents a major refactoring and enhancement of the Agent-Spec integration work originally proposed in PR #1432 ([YashaPushak:feat/agentspec](https://github.com/NVIDIA/NeMo-Agent-Toolkit/pull/1432)). This branch addresses review feedback from @willkill07 and implements a more robust, production-ready Agent-Spec plugin that follows NeMo Agent Toolkit's standard plugin architecture patterns.

## What This Branch Does

This branch implements a complete Oracle Agent-Spec integration plugin for NeMo Agent Toolkit, allowing users to run Agent-Spec YAML/JSON configurations as native NAT workflows with full observability, profiling, and evaluation capabilities.

### Key Features

1. **Agent-Spec Workflow Wrapper**: Converts Agent-Spec YAML/JSON configurations to LangGraph `CompiledStateGraph` components using `pyagentspec`'s LangGraph adapter, then wraps them as NAT Functions following the same pattern as `LanggraphWrapperFunction`.

2. **Flexible Configuration Support**:
   - File-based: `spec_file` for referencing Agent-Spec YAML/JSON files
   - Inline YAML/JSON: `spec_yaml` and `spec_json` for programmatic/dynamic configurations
   - Validator ensures exactly one source is provided

3. **NAT Tool Integration**: 
   - `tool_names` field supports referencing NAT tools via `FunctionRef` or `FunctionGroupRef`
   - Automatic tool resolution through `builder.get_tools()`
   - Merges NAT tools with Agent-Spec-defined tools (NAT tools take precedence)

4. **Component Reuse Support**:
   - `components_registry` for manually mapping component IDs to NAT-managed components (e.g., LLMs)
   - Documented limitations and best practices for component reuse with embedded Agent-Spec configs

5. **Conversation History Management**:
   - `max_history` field for trimming message history before execution
   - Uses `langchain_core.messages.trim_messages` to limit context size

6. **Auto-detection Features**:
   - Automatically detects `ClientTool` usage and configures `MemorySaver` checkpointer when needed

### Architecture

This implementation follows NeMo Agent Toolkit's standard plugin architecture:

- **Plugin Registration**: Uses `@register_function` decorator with proper framework wrappers (`LLMFrameworkEnum.LANGCHAIN`)
- **Function Wrapper Pattern**: `AgentSpecWrapperFunction` follows the same pattern as `LanggraphWrapperFunction`, handling input/output conversion and delegating execution to the underlying LangGraph
- **Separate Package**: Implemented as `nvidia-nat-agent-spec` package with proper dependencies (`nvidia-nat-core`, `pyagentspec[langgraph,langgraph_mcp]`)
- **Type Safety**: Full Pydantic models for configuration (`AgentSpecWrapperConfig`) and input/output (`AgentSpecWrapperInput`, `AgentSpecWrapperOutput`)
- **Error Handling**: Comprehensive error handling with clear error messages

### Changes from Original PR #1432

This branch addresses several architectural concerns raised in the review:

1. **Enhanced E2E Test Validation**: Added comprehensive validation in smoke tests (graph type, message types, content checks)
2. **Type Annotations**: Added explicit return type annotation (`-> Self`) to `_validate_sources` validator
3. **Documentation**: Documented component reuse limitations and best practices for embedded Agent-Spec configurations
4. **Package Structure**: Maintains separate package structure (not an optional dependency group)
5. **Code Organization**: Implementation details properly separated from configuration models

### Testing

- **55 unit tests** covering:
  - Configuration validation (file, YAML, JSON sources)
  - Tool integration (`tool_names`, `tool_registry`)
  - Component registry functionality
  - Message trimming (`max_history`)
  - Error handling and edge cases
  
- **2 end-to-end integration tests** (smoke tests):
  - File-based Agent-Spec YAML execution
  - Inline YAML execution
  - Tests skip gracefully without `NVIDIA_API_KEY`

### Usage Example

functions:
  agent_spec_workflow:
    _type: agent_spec_wrapper
    spec_file: path/to/agent_spec.yaml
    tool_names:
      - FunctionRef("wiki_search")
      - FunctionGroupRef("search_tools")
    max_history: 20
    description: "Agent-Spec workflow with NAT tool integration"
    
    
    
     ## Important: Dependency Conflicts
   
   The `agent-spec` extra cannot be installed alongside `langchain`, `most`, or `vanna` extras due to incompatible `langchain-core`/`langgraph` version requirements:
   - `pyagentspec` requires: `langchain-core<1.0.0, langgraph<1.0.0`
   - `nvidia-nat-langchain` requires: `langchain-core>=1.2.6, langgraph>=1.0.5`
   
   Users must choose one or the other, or use separate environments.

   ## Lock File Updates
   
   The `uv.lock` file has been updated to include dependencies for the new `nvidia-nat-agent-spec` package. This is expected and necessary when adding a new package.

### CODE REVIEW

File: packages/nvidia_nat_agent_spec/src/nat/plugins/agent_spec/agent_spec_workflow.py [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+    if config.tool_names:
+        try:
+            tools = await builder.get_tools(tool_names=config.tool_names, wrapper_type=LLMFrameworkEnum.LANGCHAIN)
+            if tools:
+                tool_registry = {tool.name: tool for tool in tools}
+        except Exception as e:
+            logger.warning(f"Failed to resolve tools from tool_names: {e}")
+
+    if config.tool_registry:
+        tool_registry.update(config.tool_registry)
```
Comment (willkill07) [inline packages/nvidia_nat_agent_spec/src/nat/plugins/agent_spec/agent_spec_workflow.py:345]: We should raise an error on any intersection of tool names rather than blindly take precedence

File: packages/nvidia_nat_agent_spec/src/nat/plugins/agent_spec/agent_spec_workflow.py [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+        assert config.spec_yaml is not None  # Type narrowing: validator ensures this is set
+        spec_content = config.spec_yaml
+        spec_format = "yaml"
+        source_description = "inline YAML"
+    else:
+        # Use inline JSON (config.spec_json)
+        assert config.spec_json is not None  # Type narrowing: validator ensures this is set
+        spec_content = config.spec_json
+        spec_format = "json"
+        source_description = "inline JSON"
```
Comment (willkill07) [inline packages/nvidia_nat_agent_spec/src/nat/plugins/agent_spec/agent_spec_workflow.py:309]: Hoist as a `_load_spec` function?

File: packages/nvidia_nat_agent_spec/src/nat/plugins/agent_spec/agent_spec_workflow.py [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+    """
+    try:
+        from pyagentspec.adapters.langgraph import AgentSpecLoader
+    except ImportError as e:
+        raise ImportError(
+            "Failed to import pyagentspec.adapters.langgraph. "
+            "Install pyagentspec with langgraph extras:\n"
+            "  uv pip install \"pyagentspec[langgraph,langgraph_mcp]>=26.1.0\"\n\n"
+            "Note: The root pyproject.toml includes override-dependencies to resolve "
+            "langchain-core version conflicts between pyagentspec and nvidia-nat-langchain."
```
Comment (willkill07) [inline packages/nvidia_nat_agent_spec/src/nat/plugins/agent_spec/agent_spec_workflow.py:281]: This should be a message to install `nvidia-nat[agent_spec]`

File: packages/nvidia_nat_agent_spec/pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+documentation = "https://docs.nvidia.com/nemo/agent-toolkit/latest/"
+source = "https://github.com/NVIDIA/NeMo-Agent-Toolkit"
+
+[tool.setuptools_dynamic_dependencies]
+dependencies = [
+  # Keep package version constraints as open as possible to avoid conflicts with other packages. Always define a minimum
+  # version when adding a new package. If unsure, default to using `~=` instead of `==`. Does not apply to nvidia-nat packages.
+  # Keep sorted!!!
+  "nvidia-nat-core == {version}",
+  "pyagentspec[langgraph,langgraph_mcp]>=26.1.0",
```
Comment (willkill07) [inline packages/nvidia_nat_agent_spec/pyproject.toml:43]: Is unbounded upper-bound really safe? I'd rather see: `~=26.1` (>=26.1,<27.0)

File: pyproject.toml [hunk 5]
```diff
@@ -199,12 +152,20 @@ conflicts = [
     [{ extra = "adk" }, { extra = "ragaai" }],
     [{ extra = "strands" }, { extra = "ragaai" }],
     [{ extra = "most" }, { extra = "ragaai" }],
+    # agent-spec<>langchain have incompatible langchain-core/langgraph versions
+    # pyagentspec (used by agent-spec) requires langchain-core<1.0.0, langgraph<1.0.0
+    # nvidia-nat-langchain requires langchain-core>=1.2.6, langgraph>=1.0.5
+    # These version ranges are incompatible and cannot coexist in the same workspace
+    # Users must choose one or the other, or install agent-spec in a separate environment
```
Comment (willkill07) [inline pyproject.toml:159]: remove unnecessary wording
```suggestion
```

File: pyproject.toml
Comment (willkill07) [inline pyproject.toml:1]: You should also generate a `uv.lock` in this directory

File: packages/nvidia_nat_agent_spec/tests/conftest.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+    # If NVIDIA_API_KEY is set but OPENAI_API_KEY is not, set it
+    if nvidia_key and not openai_key:
+        os.environ["OPENAI_API_KEY"] = nvidia_key
+        yield
+        # Clean up: only remove if we set it
+        if "OPENAI_API_KEY" in os.environ:
+            del os.environ["OPENAI_API_KEY"]
+    else:
+        yield
```
Comment (willkill07) [inline packages/nvidia_nat_agent_spec/tests/conftest.py:50]: This has the potential to be extremely confusing and dangerous. There has to be a better way.

At minimum:
- do not autouse
- give a name such as `map_nvidia_api_key_to_openai_api_key`

File: packages/nvidia_nat_agent_spec/tests/test_nim_integration.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+
+from nat.builder.workflow_builder import WorkflowBuilder
+from nat.plugins.agent_spec.agent_spec_workflow import AgentSpecWrapperConfig, register
+
+
+@pytest.mark.integration
+@pytest.mark.skipif(
+    not os.getenv("NVIDIA_API_KEY"),
+    reason="Requires NVIDIA_API_KEY environment variable"
+)
```
Comment (willkill07) [inline packages/nvidia_nat_agent_spec/tests/test_nim_integration.py:31]: We already have an `nvidia_api_key` fixture that can be used -- it is part of `nvidia_nat_test`

File: packages/nvidia_nat_agent_spec/tests/test_e2e_nat_workflow.py [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+from pathlib import Path
+
+import pytest
+
+
+@pytest.mark.integration
+@pytest.mark.skipif(
+    not os.getenv("NVIDIA_API_KEY"),
+    reason="Requires NVIDIA_API_KEY environment variable"
+)
```
Comment (willkill07) [inline packages/nvidia_nat_agent_spec/tests/test_e2e_nat_workflow.py:40]: Prefer `nvidia_api_key` fixture as parameter

File: packages/nvidia_nat_agent_spec/tests/test_e2e_nat_workflow.py [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+                pytest.skip(f"Invocation failed (may need network/API): {e}")
+
+
+
+
+@pytest.mark.integration
+@pytest.mark.skipif(
+    not os.getenv("NVIDIA_API_KEY"),
+    reason="Requires NVIDIA_API_KEY environment variable"
+)
```
Comment (willkill07) [inline packages/nvidia_nat_agent_spec/tests/test_e2e_nat_workflow.py:122]: Prefer `nvidia_api_key` fixture as parameter

File: packages/nvidia_nat_agent_spec/tests/test_nim_integration.py [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+                print(f"✓ Successfully loaded Agent-Spec YAML: {test_yaml}")
+                print(f"✓ Graph type: {type(wrapper_function._graph)}")
+                print(f"✓ Graph compiled successfully")
+
+
+@pytest.mark.integration
+@pytest.mark.skipif(
+    not os.getenv("NVIDIA_API_KEY"),
+    reason="Requires NVIDIA_API_KEY environment variable"
+)
```
Comment (willkill07) [inline packages/nvidia_nat_agent_spec/tests/test_nim_integration.py:77]: Prefer `nvidia_api_key` fixture as parameter

File: packages/nvidia_nat_agent_spec/tests/fixtures/weather_agent_with_tool.yaml [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+
+component_type: Agent
+description: 'Weather agent with a client tool'
+name: weather_agent
+inputs: []
+llm_config:
+  component_type: OpenAiCompatibleConfig
+  name: "NIM LLM"
+  model_id: "meta/llama-3.1-8b-instruct"
+  # NIM endpoint - will use NVIDIA_API_KEY from environment
```
Comment (willkill07) [inline packages/nvidia_nat_agent_spec/tests/fixtures/weather_agent_with_tool.yaml:12]: Is this true? because it seems like you have to set `OPENAI_API_KEY`

File: packages/nvidia_nat_agent_spec/tests/fixtures/weather_agent_with_tool.yaml [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+  url: "https://integrate.api.nvidia.com/v1"
```
Comment (willkill07) [inline packages/nvidia_nat_agent_spec/tests/fixtures/weather_agent_with_tool.yaml:13]: Just a note on how messed up their implementation is...

https://github.com/oracle/agent-spec/blob/main/pyagentspec/src/pyagentspec/llms/openaicompatibleconfig.py#L29

`OpenAiCompatibleConfig` has an optional `api_key`.

But ironically, they don't even map it!

https://github.com/oracle/agent-spec/blob/main/pyagentspec/src/pyagentspec/adapters/langgraph/_langgraphconverter.py#L1233-L1242

File: packages/nvidia_nat_agent_spec/src/nat/plugins/agent_spec/agent_spec_workflow.py [hunk 13]
```diff
@@ -0,0 +1,371 @@
...
+
+    model_config = ConfigDict(extra="forbid")
+
+    description: str = Field(default="", description="Description of the Agent-Spec workflow")
+    spec_file: FilePath | None = Field(
+        default=None,
+        description="Path to the Agent-Spec YAML/JSON configuration file. "
+        "Recommended for NAT YAML workflow configurations."
+    )
+    spec_yaml: str | None = Field(
```
Comment (jschweiz) [inline packages/nvidia_nat_agent_spec/src/nat/plugins/agent_spec/agent_spec_workflow.py:102]: here, the workflow config object contains only a stringified version of an agentspec config. This means that when using the optimizer, the optimizer cannot navigate in this config to discover parameters that can be optimized nor the values/ranges they should be optimized on.

I see two ways to make this work.

# Using a dynamic basemodel

This is what I'm currently experimenting with. I basically added 2 attributes to this workflow config, one "dynamic_optimizeable_parameters" that's empty initially, and a "optimizeable_parameters_dict" dict.
It looks like this in the config:
```
workflow:
  _type: agent_spec_wrapper
  spec_file: stage_2.yaml

  # this is required for adding dynamically the optimizable parameters in the config object
  optimizable_params:
    - dynamic_optimizable_parameters

  # dict for optimize-able parameters with values
  optimizable_parameters_dict:
    temperature_ref:
      range:
        low: 0.0
        high: 1.0
        step: 0.1
    max_tokens_ref:
      range:
        low: 2048
        high: 8096
        step: 2048
```
Then, as a post-validation of the `AgentSpecWrapperConfig` object, I create a dynamic base-model:
```
@model_validator(mode="after")
    def _build_prompts_opt(self) -> Self:
        fields: dict[str, tuple[type, Any]] = {}
        for var_name, var in self.optimizable_parameters_dict.items():
            if var.values is not None:
                fields[var_name] = (
                    dict,
                    OptimizableField(
                        default=var.values[0],
                        space=SearchSpace(values=var.values),
                    ),
                )
            elif var.range is not None:
                fields[var_name] = (
                    dict,
                    OptimizableField(
                        default=var.range.low,
                        space=SearchSpace(low=var.range.low, high=var.range.high, step=var.range.step),
                    ),
                )
            else:
                raise ValueError('Not supported')

        DynamicPrompts = create_model("DynamicPrompts", **fields)
        self.dynamic_optimizable = DynamicPrompts()
        return self
```

This allows the optimizer to still see all the parameters inside the config, even if these attributes are not decorated with OptimizableField.


# Cleaner approach: patch our config objects with the "optimizable" annotation

We also represent config objects as base models, so one way would be to extend the loading of the workflow config with our own loading logic, and patch our base models with the optimizable field annotator, for example on `temperature` like in https://github.com/oracle/agent-spec/blob/main/pyagentspec/src/pyagentspec/llms/llmgenerationconfig.py#L23.

This would allow a more seemless and native use of the optimizer with agentspec configs.


Do you see another potential solution, and what do you think about these?

File: packages/nvidia_nat_agent_spec/src/nat/plugins/agent_spec/agent_spec_workflow.py [hunk 14]
```diff
@@ -0,0 +1,371 @@
...
+        "To use this, ensure your Agent-Spec YAML includes an 'id' field in the LLM config, "
+        "then map that ID to your NAT LLM: components_registry={'llm_id': nat_llm}.",
+    )
+    checkpointer: Any | None = Field(
+        default=None,
+        description="Optional LangGraph checkpointer (e.g., MemorySaver). Required when using "
+        "ClientTool in Agent-Spec YAML. If not provided and ClientTool is detected, "
+        "MemorySaver will be used automatically.",
+    )
+    max_history: int = Field(
```
Comment (jschweiz) [inline packages/nvidia_nat_agent_spec/src/nat/plugins/agent_spec/agent_spec_workflow.py:144]: we support this concept natively in agentspec, I don't think we need this here (we can remove the part about pruning the history before the invoke/stream calls too)

File: packages/nvidia_nat_agent_spec/src/nat/plugins/agent_spec/agent_spec_workflow.py [hunk 15]
```diff
@@ -0,0 +1,371 @@
...
+from nat.builder.function import Function
+from nat.cli.register_workflow import register_function
+from nat.data_models.component_ref import FunctionGroupRef
+from nat.data_models.component_ref import FunctionRef
+from nat.data_models.function import FunctionBaseConfig
+
+logger = logging.getLogger(__name__)
+
+
+class AgentSpecWrapperInput(BaseModel):
```
Comment (jschweiz) [inline packages/nvidia_nat_agent_spec/src/nat/plugins/agent_spec/agent_spec_workflow.py:58]: we would need to extend to be also a dict of inputs, since soem agentspec config takes inputs and not a list of messages.

File: packages/nvidia_nat_agent_spec/src/nat/plugins/agent_spec/agent_spec_workflow.py [hunk 16]
```diff
@@ -0,0 +1,371 @@
...
+    """Input model for the Agent-Spec wrapper."""
+
+    model_config = ConfigDict(extra="allow")
+
+    messages: list[BaseMessage]
+
+
+class AgentSpecWrapperOutput(BaseModel):
```
Comment (jschweiz) [inline packages/nvidia_nat_agent_spec/src/nat/plugins/agent_spec/agent_spec_workflow.py:66]: similarly, we would need to support outputting a dict of outputs. Do you already support handling such input/output dictionnaries, and how does it plain with the evaluator (the mappings `question` -> `WrapperInput` and `WrapperOutput` -> `output` are defined below in `AgentSpecWrapperFunction`, is that what we always want?
